### PR TITLE
[Fix] `destructuring-assignment`: Handle destructuring of useContext in SFC

### LINKED ISSUE #2309: [react/destructuring-assignment] Shows weird behaviour when I declare a variable with the name context
In a functional component I get an error when I declare a variable with the name context.
Example code:

```javascript
const Test = props => {
  const context = useContext(someContext);

  return (
    <div {...props}>
      {context.test} // Here message: 'Must use destructuring context'
    </div>
  );
}
```

### PR DESCRIPTION
**Pull request by Zinyon**
i reopen PR on resolve issue - #2309 cause i did a lot of commits while editing.

* add rule of handling destructuring of useContext in SFC
* delete detecting variable named 'context' in SFC

Fixes #2309. Closes #2787.

### CODE REVIEW

File: lib/rules/destructuring-assignment.js [hunk 1]
```diff
@@ -61,14 +62,24 @@ module.exports = {
     }
 
     function handleSFCUsage(node) {
-      // props.aProp || context.aProp
-      const isPropUsed = (node.object.name === 'props' || node.object.name === 'context') && !isAssignmentLHS(node);
+      // props.aProp
+      const isPropUsed = (node.object.name === 'props') && !isAssignmentLHS(node);
```
Comment (ljharb) [inline lib/rules/destructuring-assignment.js:66]:
```suggestion
      const isPropUsed = node.object.name === 'props' && !isAssignmentLHS(node);
```

File: lib/rules/destructuring-assignment.js [hunk 2]
```diff
@@ -125,13 +136,29 @@ module.exports = {
...
 
         const destructuring = (node.init && node.id && node.id.type === 'ObjectPattern');
+        const identifier = (node.init && node.id && node.id.type === 'Identifier');
         // let {foo} = props;
-        const destructuringSFC = destructuring && (node.init.name === 'props' || node.init.name === 'context');
+        const destructuringSFC = destructuring && node.init.name === 'props';
+        // let {foo} = useContext(aContext);
+        const destructuringUseContext = destructuring && node.init.callee && node.init.callee.name === 'useContext';
+        // let foo = useContext(aContext);
+        const assignUseContext = identifier && node.init.callee && node.init.callee.name === 'useContext';
```
Comment (ljharb) [inline lib/rules/destructuring-assignment.js:145]: are we sure `useContext` is the react hook? I think we should make sure that nobody's using a custom hook or function named `useContext`, and that this `useContext` is imported/required from `react`.

**Review by ljharb — CHANGES_REQUESTED** [review]

File: lib/rules/destructuring-assignment.js (hunk 2 above)
Comment (Zinyon) [reply lib/rules/destructuring-assignment.js:145]: https://reactjs.org/docs/hooks-reference.html#usecontext
here in official docs of react says ``useContext`` is one of basic hooks with ``useState``, ``useEffect``.
and we import ``useContext`` like ``import {useContext} from 'react'``;

I think if somebody's using function named ``useContext`` which overwrite basic hooks , I think we should make another rule about custom hooks which overwrite basic hooks.

Comment (ljharb) [reply lib/rules/destructuring-assignment.js:145]: Someone might be using a version of React that doesn't have hooks.

Comment (Zinyon) [reply lib/rules/destructuring-assignment.js:145]: then is there any possible way to check version of react? or we should not include this part.

Comment (ljharb) [reply lib/rules/destructuring-assignment.js:145]: Yes! Check out `lib/util/version.js`'s `testReactVersion` method.

Comment (Zinyon) [reply lib/rules/destructuring-assignment.js:145]: do I have to import ``testReactVersion`` method in ``lib/rules/destructuring-assignment'`` to check react version in rules? I have no idea.. could you help me with this?

Comment (ljharb) [reply lib/rules/destructuring-assignment.js:145]: Yes, that's right

**Comment by ljharb** [comment]
@Zinyon are you still interested in completing this PR?

**Comment by ljharb** [comment]
The `isDestructuredFromPragmaImport` helper might be useful here.

**Comment by Zinyon** [comment]
@ljharb sure, i will look around ``isDestructuredFromPragmaImport`` helper and complete PR soon!

**Comment by ljharb** [comment]
I went ahead and added the react version dependence; i'll leave "make sure it's destructured from the pragma" as a followup.

**Review by ljharb — APPROVED** [review]

**Pull request merged**